src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreUnconfiguredInputDataSource.cs

```diff
@@ -38,37 +38,26 @@ protected override IDisposable LinkExternalInput(ITargetBlock<RestoreInfo> targe
-                projects => projects.Select(project => project.Services.ExportProvider.GetExportedValueOrDefault<IPackageRestoreConfiguredInputDataSource>())
-                                    .WhereNotNull() // Filter out those without PackageReference
-                                    .Select(DropConfiguredProjectVersions),
-                includeSourceVersions: true);
-
-            disposables.Add(restoreConfiguredInputSource);
-
-            IProjectValueDataSource<IConfigurationGroup<ConfiguredProject>> activeConfiguredProjectsSource = _activeConfigurationGroupService.ActiveConfiguredProjectGroupSource;
-            disposables.Add(activeConfiguredProjectsSource.SourceBlock.LinkTo(restoreConfiguredInputSource, DataflowOption.PropagateCompletion));
-
-            // Dataflow from two configurations can depend on a same unconfigured level data source, and processes it at a different speed.
-            // Introduce a forward-only block to prevent regressions in versions.
-            var forwardOnlyBlock = ProjectDataSources.CreateDataSourceVersionForwardOnlyFilteringBlock<IReadOnlyCollection<PackageRestoreConfiguredInput>>();
-            disposables.Add(restoreConfiguredInputSource.SourceBlock.LinkTo(forwardOnlyBlock, DataflowOption.PropagateCompletion));
+            var joinBlock = new ConfiguredProjectDataSourceJoinBlock<PackageRestoreConfiguredInput>(
+                project => project.Services.ExportProvider.GetExportedValueOrDefault<IPackageRestoreConfiguredInputDataSource>(),
+                JoinableFactory,
+                _project);
 
             // Transform all restore data -> combined restore data
-            DisposableValue<ISourceBlock<RestoreInfo>> mergeBlock = forwardOnlyBlock.TransformWithNoDelta(update => update.Derive(MergeRestoreInputs));
-            disposables.Add(mergeBlock);
+            DisposableValue<ISourceBlock<RestoreInfo>> mergeBlock = joinBlock.TransformWithNoDelta(update => update.Derive(MergeRestoreInputs));
 
-            // Set the link up so that we publish changes to target block
-            mergeBlock.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
+            return new DisposableBag
+            {
+                joinBlock,
+                mergeBlock,
 
-            // Join the source blocks, so if they need to switch to UI thread to complete 
-            // and someone is blocked on us on the same thread, the call proceeds
-            JoinUpstreamDataSources(restoreConfiguredInputSource, activeConfiguredProjectsSource);
+                // Link the active configured projects to our join block
+                _activeConfigurationGroupService.ActiveConfiguredProjectGroupSource.SourceBlock.LinkTo(joinBlock, DataflowOption.PropagateCompletion),
 
-            return disposables;
+                // Set the link up so that we publish changes to target block
+                mergeBlock.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion)
```

Comment:
usually, we don't need dispose those links other than the joinBlock & the top link to the ProjectGroupSource, because it will release mergeBlock block from memory (it is only held by the its source block). of course, it doesn't hurt much by holding the link in the disposable bug.

Reply:
@lifengl I don't see why the merge block could be freed. Isn't it participating in the data flow?